Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUBufferImpl.cpp

```diff
@@ -45,11 +45,20 @@ BufferImpl::~BufferImpl()
     wgpuBufferRelease(m_backing);
 }
 
+static Size64 getMappedSize(WGPUBuffer backing, std::optional<Size64> size, Size64 offset)
+{
+    if (size.has_value())
+        return size.value();
+
+    auto bufferSize = wgpuBufferGetSize(backing);
+    return bufferSize > offset ? (bufferSize - offset) : 0;
+}
+
 void BufferImpl::mapAsync(MapModeFlags mapModeFlags, Size64 offset, std::optional<Size64> size, CompletionHandler<void()>&& callback)
 {
     auto backingMapModeFlags = m_convertToBackingContext->convertMapModeFlagsToBacking(mapModeFlags);
 
-    auto usedSize = size.value_or(WGPU_WHOLE_MAP_SIZE);
+    auto usedSize = getMappedSize(m_backing, size, offset);
 
```

Comment:
I feel like we're missing a bunch of validation here...

Reply:
The validation happens at the WebGPU.framework layer